File: dbflow/src/main/java/com/raizlabs/android/dbflow/structure/RetrievalAdapter.java
package com.raizlabs.android.dbflow.structure;

import android.database.Cursor;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.raizlabs.android.dbflow.config.DatabaseConfig;
import com.raizlabs.android.dbflow.config.DatabaseDefinition;
import com.raizlabs.android.dbflow.config.FlowManager;
import com.raizlabs.android.dbflow.config.TableConfig;
import com.raizlabs.android.dbflow.sql.language.OperatorGroup;
import com.raizlabs.android.dbflow.sql.language.SQLite;
import com.raizlabs.android.dbflow.sql.queriable.ListModelLoader;
import com.raizlabs.android.dbflow.sql.queriable.SingleModelLoader;
import com.raizlabs.android.dbflow.structure.database.DatabaseWrapper;
import com.raizlabs.android.dbflow.structure.database.FlowCursor;

/**
 * Description: Provides a base retrieval class for all {@link Model} backed
 * adapters.
 */
@SuppressWarnings("NullableProblems")
public abstract class RetrievalAdapter<TModel> {

    private SingleModelLoader<TModel> singleModelLoader;
    private ListModelLoader<TModel> listModelLoader;

    private TableConfig<TModel> tableConfig;

    public RetrievalAdapter(@NonNull DatabaseDefinition databaseDefinition) {
        DatabaseConfig databaseConfig = FlowManager.getConfig()
                .getConfigForDatabase(databaseDefinition.getAssociatedDatabaseClassFile());
        if (databaseConfig != null) {
            tableConfig = databaseConfig.getTableConfigForTable(getModelClass());
            if (tableConfig != null) {
                if (tableConfig.getSingleModelLoader() != null) {
                    singleModelLoader = tableConfig.getSingleModelLoader();
                }

                if (tableConfig.getListModelLoader() != null) {
                    listModelLoader = tableConfig.getListModelLoader();
                }
            }
        }
    }

    /**
     * Force loads the model from the DB. Even if caching is enabled it will requery the object.
     */
    public void load(@NonNull TModel model) {
        load(model, FlowManager.getDatabaseForTable(getModelClass()).getWritableDatabase());
    }

    /**
     * Force loads the model from the DB. Even if caching is enabled it will requery the object.
     */
    public void load(@NonNull TModel model, DatabaseWrapper databaseWrapper) {
        getNonCacheableSingleModelLoader().load(databaseWrapper,
                SQLite.INSTANCE.select()
                        .from(getModelClass())
                        .where(getPrimaryConditionClause(model)).getQuery(),
                model);
    }

    /**
     * Assigns the {@link Cursor} data into the specified {@link TModel}
     *
     * @param model  The model to assign cursor data to
     * @param cursor The cursor to load into the model
     */
    public abstract void loadFromCursor(@NonNull FlowCursor cursor, @NonNull TModel model);

    /**
     * @param model The model to query values from
     * @return True if it exists as a row in the corresponding database table
     */
    public boolean exists(@NonNull TModel model) {
        return exists(model, FlowManager.getDatabaseForTable(getModelClass()).getWritableDatabase());
    }

    /**
     * @param model The model to query values from
     * @return True if it exists as a row in the corresponding database table
     */
    public abstract boolean exists(@NonNull TModel model,
                                   @NonNull DatabaseWrapper databaseWrapper);

    /**
     * @param model The primary condition clause.
     * @return The clause that contains necessary primary conditions for this table.
     */
    public abstract OperatorGroup getPrimaryConditionClause(@NonNull TModel model);

    /**
     * @return the model class this adapter corresponds to
     */
    @NonNull
    public abstract Class<TModel> getModelClass();

    @Nullable
    protected TableConfig<TModel> getTableConfig() {
        return tableConfig;
    }

    /**
     * @return A new {@link ListModelLoader}, caching will override this loader instance.
     */
    @NonNull
    public ListModelLoader<TModel> getListModelLoader() {
        if (listModelLoader == null) {
            listModelLoader = createListModelLoader();
        }
        return listModelLoader;
    }

    /**
     * @return A new {@link ListModelLoader}, caching will override this loader instance.
     */
    @NonNull
    protected ListModelLoader<TModel> createListModelLoader() {
        return new ListModelLoader<>(getModelClass());
    }

    /**
     * @return A new {@link SingleModelLoader}, caching will override this loader instance.
     */
    @NonNull
    protected SingleModelLoader<TModel> createSingleModelLoader() {
        return new SingleModelLoader<>(getModelClass());
    }

    @NonNull
    public SingleModelLoader<TModel> getSingleModelLoader() {
        if (singleModelLoader == null) {
            singleModelLoader = createSingleModelLoader();
        }
        return singleModelLoader;
    }

    /**
     * @return A new instance of a {@link SingleModelLoader}. Subsequent calls do not cache
     * this object so it's recommended only calling this in bulk if possible.
     */
    @NonNull
    public SingleModelLoader<TModel> getNonCacheableSingleModelLoader() {
        return new SingleModelLoader<>(getModelClass());
    }

    /**
     * @return A new instance of a {@link ListModelLoader}. Subsequent calls do not cache
     * this object so it's recommended only calling this in bulk if possible.
     */
    @NonNull
    public ListModelLoader<TModel> getNonCacheableListModelLoader() {
        return new ListModelLoader<>(getModelClass());
    }

    /**
     * Overrides the default implementation and allows you to provide your own implementation. Defines
     * how a single {@link TModel} is loaded.
     *
     * @param singleModelLoader The loader to use.
     */
    public void setSingleModelLoader(@NonNull SingleModelLoader<TModel> singleModelLoader) {
        this.singleModelLoader = singleModelLoader;
    }

    /**
     * Overrides the default implementation and allows you to provide your own implementation. Defines
     * how a list of {@link TModel} are loaded.
     *
     * @param listModelLoader The loader to use.
     */
    public void setListModelLoader(@NonNull ListModelLoader<TModel> listModelLoader) {
        this.listModelLoader = listModelLoader;
    }

}

File: dbflow/src/main/java/com/raizlabs/android/dbflow/structure/RetrievalAdapter.kt
package com.raizlabs.android.dbflow.structure

import android.database.Cursor
import com.raizlabs.android.dbflow.config.DatabaseDefinition
import com.raizlabs.android.dbflow.config.FlowManager
import com.raizlabs.android.dbflow.config.TableConfig
import com.raizlabs.android.dbflow.sql.language.OperatorGroup
import com.raizlabs.android.dbflow.sql.language.select
import com.raizlabs.android.dbflow.sql.language.where
import com.raizlabs.android.dbflow.sql.queriable.ListModelLoader
import com.raizlabs.android.dbflow.sql.queriable.SingleModelLoader
import com.raizlabs.android.dbflow.structure.database.DatabaseWrapper
import com.raizlabs.android.dbflow.structure.database.FlowCursor

/**
 * Description: Provides a base retrieval class for all [Model] backed
 * adapters.
 */
abstract class RetrievalAdapter<T : Any>(databaseDefinition: DatabaseDefinition) {

    /**
     * Overrides the default implementation and allows you to provide your own implementation. Defines
     * how a single [T] is loaded.
     *
     * @param singleModelLoader The loader to use.
     */
    private var _singleModelLoader: SingleModelLoader<T>? = null

    var singleModelLoader: SingleModelLoader<T>
        get() {
            if (_singleModelLoader == null) {
                _singleModelLoader = createSingleModelLoader()
            }
            return _singleModelLoader!!
        }
        set(value) {
            this._singleModelLoader = value
        }
    /**
     * @return A new [ListModelLoader], caching will override this loader instance.
     */
    /**
     * Overrides the default implementation and allows you to provide your own implementation. Defines
     * how a list of [T] are loaded.
     *
     * @param listModelLoader The loader to use.
     */
    private var _listModelLoader: ListModelLoader<T>? = null

    var listModelLoader: ListModelLoader<T>
        get() {
            if (_listModelLoader == null) {
                _listModelLoader = createListModelLoader()
            }
            return _listModelLoader!!
        }
        set(value) {
            this._listModelLoader = value
        }

    protected var tableConfig: TableConfig<T>? = null
        private set

    /**
     * @return the model class this adapter corresponds to
     */
    abstract val modelClass: Class<T>

    /**
     * @return A new instance of a [SingleModelLoader]. Subsequent calls do not cache
     * this object so it's recommended only calling this in bulk if possible.
     */
    val nonCacheableSingleModelLoader: SingleModelLoader<T>
        get() = SingleModelLoader(modelClass)

    /**
     * @return A new instance of a [ListModelLoader]. Subsequent calls do not cache
     * this object so it's recommended only calling this in bulk if possible.
     */
    val nonCacheableListModelLoader: ListModelLoader<T>
        get() = ListModelLoader(modelClass)

    init {
        val databaseConfig = FlowManager.getConfig()
                .getConfigForDatabase(databaseDefinition.associatedDatabaseClassFile)
        if (databaseConfig != null) {
            tableConfig = databaseConfig.getTableConfigForTable(modelClass)
            if (tableConfig != null) {
                tableConfig?.singleModelLoader?.let { _singleModelLoader = it }
                tableConfig?.listModelLoader?.let { _listModelLoader = it }
            }
        }
    }

    /**
     * Force loads the model from the DB. Even if caching is enabled it will requery the object.
     */
    open fun load(model: T, databaseWrapper: DatabaseWrapper) {
        nonCacheableSingleModelLoader.load(databaseWrapper,
                (databaseWrapper.select
                        from modelClass
                        where getPrimaryConditionClause(model)).query,
                model)
    }

    /**
     * Assigns the [Cursor] data into the specified [T]
     *
     * @param model  The model to assign cursor data to
     * @param cursor The cursor to load into the model
     */
    abstract fun loadFromCursor(cursor: FlowCursor, model: T, wrapper: DatabaseWrapper)

    /**
     * @param model The model to query values from
     * @return True if it exists as a row in the corresponding database table
     */
    open fun exists(model: T): Boolean =
            exists(model, FlowManager.getDatabaseForTable(modelClass).writableDatabase)

    /**
     * @param model The model to query values from
     * @return True if it exists as a row in the corresponding database table
     */
    abstract fun exists(model: T, databaseWrapper: DatabaseWrapper): Boolean

    /**
     * @param model The primary condition clause.
     * @return The clause that contains necessary primary conditions for this table.
     */
    abstract fun getPrimaryConditionClause(model: T): OperatorGroup

    /**
     * @return A new [ListModelLoader], caching will override this loader instance.
     */
    protected open fun createListModelLoader(): ListModelLoader<T> = ListModelLoader(modelClass)

    /**
     * @return A new [SingleModelLoader], caching will override this loader instance.
     */
    protected open fun createSingleModelLoader(): SingleModelLoader<T> = SingleModelLoader(modelClass)

}
/**
 * Force loads the model from the DB. Even if caching is enabled it will requery the object.
 */
